python/cudf_polars/cudf_polars/dsl/expressions/boolean.py

```diff
@@ -380,9 +380,16 @@ def do_evaluate(
             )
         elif self.name is BooleanFunction.Name.Not:
             (column,) = columns
+            # Polars semantics:
+            #   integer input: NOT => bitwise invert.
+            #   boolean input: NOT => logical NOT.
             return Column(
                 plc.unary.unary_operation(
-                    column.obj, plc.unary.UnaryOperator.NOT, stream=df.stream
+                    column.obj,
+                    plc.unary.UnaryOperator.BIT_INVERT
+                    if plc.traits.is_integral_not_bool(column.obj.type())
```

Comment:
We only need to check for `column.obj.type().id() == BOOL8` because polars will have validated that the column type is an integral type already. I think that might be clearer?

Reply:
Done in https://github.com/rapidsai/cudf/pull/20599/commits/20be6ff11c37671a464bfcf204da1dcc772829ee